v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/transformer/GenericRecordTypeConvertor.java

```diff
@@ -0,0 +1,411 @@
+/*
+ * Copyright (C) 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.transformer;
+
+import com.google.auto.value.AutoValue;
+import com.google.cloud.ByteArray;
+import com.google.cloud.Date;
+import com.google.cloud.spanner.Value;
+import com.google.cloud.teleport.v2.spanner.migrations.schema.ISchemaMapper;
+import com.google.cloud.teleport.v2.spanner.type.Type;
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.nio.ByteBuffer;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Pattern;
+import org.apache.avro.Conversions;
+import org.apache.avro.LogicalTypes;
+import org.apache.avro.Schema;
+import org.apache.avro.data.TimeConversions;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.commons.codec.binary.Hex;
+import org.apache.commons.lang3.BooleanUtils;
+import org.apache.commons.lang3.math.NumberUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Convertor Class containing methods for type conversion of various AvroTypes to Spanner {@link
+ * Value} types.
+ */
+@AutoValue
+public abstract class GenericRecordTypeConvertor {
+  private static final Logger LOG = LoggerFactory.getLogger(GenericRecordTypeConvertor.class);
+
+  public abstract ISchemaMapper iSchemaMapper();
+
+  public abstract String namespace();
+
+  public static GenericRecordTypeConvertor create(ISchemaMapper iSchemaMapper, String namespace) {
+    return new AutoValue_GenericRecordTypeConvertor(iSchemaMapper, namespace);
+  }
+
+  /**
+   * This method takes in a generic record and returns a map between the Spanner column name and the
+   * corresponding Spanner column value. This handles the data conversion logic from a GenericRecord
+   * field to a spanner Value.
+   */
+  public Map<String, Value> transformChangeEvent(GenericRecord record, String srcTableName) {
```

Comment:
From this point onward, the record contains only the data values and none of the metadata like table names (call it the "payload" record). The caller will extract the tableName/metadata_fields/payload from the "sourceRecord" (this has everything) and call this on just the "payload" record.

Why don't we take the expect sourceRecord here directly? That's because the metadata fields will have different structures for bulk and live, hence the handling for extracting the names etc will be different. The common path begins once payload is extracted out.

Reply:
Thanks for the explanation!